src/redisearch_rs/rqe_iterators/src/rqe_iterator.rs

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+use ffi::t_docId;
+use inverted_index::RSIndexResult;
+
+/// An iterator has two successful states: `OK` and `NotFound`.
+/// `OK` indicates that the iterator has found a valid result, in either "read" or "skip_to",
+/// `NotFound` indicates that a "skip_to(doc_id)" operation has found a result with a doc_id greater than the requested `doc_id`.
+pub enum RQEIteratorStatus {
+    OK(RSIndexResult),
+    NotFound(RSIndexResult),
+}
+
+/// An iterator can fail with one of the following errors:
+/// `EOF` indicates that the iterator has reached the end of the index.
+/// `TimedOut` indicates that the iterator has reached the time limit for execution.
+pub enum RQEIteratorError {
+    EOF,
+    TimedOut,
+}
+
+/// An iterator can be in one of the following states after a call to `revalidate`:
+/// `OK` indicates that the iterator is still valid and at the same position.
+/// `Moved` indicates that the iterator is still valid but lastDocID changed, and `current` is a new valid result or at EOF. If not at EOF, the `current` result should be used before the next read, or it will be overwritten.
+/// `Aborted` indicates that the iterator is no longer valid, and should not be used or rewound. Should be freed.
+pub enum RQEValidateStatus {
+    OK,
+    Moved,
+    Aborted,
+}
+
+pub trait RQEIterator {
+    /// Read the next entry from the iterator.
+    /// On a successful read, the iterator must:
+    /// 1. Set its `lastDocId` member to the new current result id
+    /// 2. Set its `current` index result property to its current result, for the caller to access if desired
+    /// returns Ok(RQEIteratorStatus::OK) with a reference to the current result or Err(RQEIteratorError) for any error.
+    fn read(&mut self) -> Result<RQEIteratorStatus, RQEIteratorError>;
```

Comment:
If my understanding of the docs is correct, this will never return `RQEIteratorStatus::NotFound`, won't it?

Reply:
Then I would recommend a different return type: `Result<Option<RSIndexResult>, TimeoutError>` (or `Option<Result<RSIndexResult, TimeoutError>>`, slightly different semantics for time outs).
The `None` case encodes the EoF scenario, as it's idiomatic for Rust iterators.
You don't have a redundant variant in the success case, so the caller code doesn't have to "discard" it.